### ISSUE TITLE
Going from notebook visualisation to internal/public dashboard

### ISSUE THREAD

**Issue by paularmand** [pyviz/panel#707]
**Is your feature request related to a problem? Please describe.**
A typical questions in a company to a data analysts / insight team encompasses the delivery of an analysis (often visually) to decision makers either realtime or day+1 via a dashboard. E.g. add a realtime visualisation to a dashboard depicting the number of visitors to our sites, or add a visualisation that shows the hourly number of visitors. These dashboards typically allow some level of drill-down.

**Describe the solution you'd like**
I would love the data analyst/scientist to be able to:
+ firstly: do the explorative work in a jupyter notebook
+ secondly: publish an output of a cell (a .servable() ) to a dashboard 

**Describe alternatives you've considered**
The data scientist currently uses an elasticsearch to store raw and processed data and have build a custom dashboarding solution (architecturally similar to  https://github.com/Smashing/smashing). This also allows to perform analysis of the raw data in kibana.

**Additional context**
+ The data scientists are typically not SW engineers and the dashboarding is just a means to an end. The responsibility of maintaining/running the dashboarding solution should be in the hands of an engineering team, but the templating and the visualisations should ideally be completely in the hands of the data scientist.
+ The workflow should be integrated in jupyterlab. In order to publish a notebook cell we could require the data scientist to add tags that reference the specifics of the dashboard template from panel.
+ Panel / bokeh should be able to serve the dashboards and run the visualisations, and hot load new visualisations to the template.
+ An additional difficulty: how to filter the dashboard and have visualisations consistent.

**Comment by jbednar** [comment]
Is there a request, a problem, or a suggestion here?  Yes, Panel should be able to work like this (except depending on what you mean by integrating in JupyterLab), if that's what you're asking.

**Comment by philippjfr** [comment]
Yes, there are multiple requests here and I'll lay them out one by one:

1. The first part of it is the ability to take the output of multiple notebooks and put them together into a single template. There are multiple approaches here but it might make sense to explore serving multiple separate sessions and then pull the different components together into the template using bokeh's ``pull_session`` functionality. I'd have to do some exploration around this.

2. The second part of this is building a JupyterLab extension which marks the output of a cell as servable and allows attaching a name/label to it so it can be referred to from an external template

3. The ability to hotload the different inputs to a template, i.e. when a notebook changes the template should pull re-execute the component and update the output in the template.

4. Have some way to query and filter notebooks and inspect their tagged outputs to make it easier to find the component you want to combine into a template.

Overall this approach will make it possible to write individual and isolated components which get combined into an overall dashboard. The main issues to solve here is to what extent the output components can be made independent of the template they will eventually be inserted into. Since the component should have the appropriate layout and styling to fit into a particular template achieving a clean separation may be difficult, but maybe we can get quite far with bokeh themes.

**Comment by jbednar** [comment]
This sounds similar to, but maybe not exactly the same as, something I've been contemplating, which is a "dashboard dashboard".  I.e., I'm imagining a group or organization whose members publish lots of separate dashboards, each maintained independently, but which each generate some KPI (numeric metric value) or small graphical result (traffic light, speedometer, etc.) that can be used for a thumbnail that could appear in a separate higher-level dashboard.  The higher-level "dashboard of dashboards" would show an array of such thumbnails, and people would click on ones that seemed like they need investigating, e.g. because a KPI or traffic light indicates some problem that needs checking out.  

To do this, I was imagining that the dashboard dashboard would periodically query some REST API on each of the other dashboards, collecting the results into the higher-level display and letting people drill down into the underlying dashboard where needed.  I was _not_ thinking of the underlying dashboard as being able to publish some arbitrary plot to this interface, but if it did, then maybe your proposal and mine could amount to the same thing?  I also wasn't personally thinking of this as having to do with JupyterLab, because I'm not expecting the consumers of the dashboard dashboard to be Python-savvy necessarily, but that's a separable issue.

**Comment by paularmand** [comment]
Are there security considerations, we need to take into account going along? See also https://docs.bokeh.org/en/latest/docs/user_guide/server.html#shared-publishing

**Comment by paularmand** [comment]
@philippjfr is this related to the lumen project?
